src/Servers/Kestrel/Core/src/Internal/Http/Http1UpgradeMessageBody.cs

```diff
@@ -69,12 +72,49 @@ public override Task StopAsync()
 
         public override bool TryReadInternal(out ReadResult readResult)
         {
-            return _context.Input.TryRead(out readResult);
+            // Ignore the canceled readResult unless it was canceled by the user.
+            do
+            {
+                if (!_context.Input.TryRead(out readResult))
+                {
+                    return false;
+                }
+            } while (readResult.IsCanceled && Interlocked.Exchange(ref _userCanceled, 0) == 0);
+
+            return true;
         }
 
         public override ValueTask<ReadResult> ReadAsyncInternal(CancellationToken cancellationToken = default)
         {
-            return _context.Input.ReadAsync(cancellationToken);
+            ReadResult readResult;
+
+            // Ignore the canceled readResult unless it was canceled by the user.
+            do
+            {
+                var readTask = _context.Input.ReadAsync(cancellationToken);
+
+                if (!readTask.IsCompletedSuccessfully)
+                {
+                    return ReadAsyncInternalAwaited(readTask, cancellationToken);
+                }
+
+                readResult = readTask.GetAwaiter().GetResult();
+            } while (readResult.IsCanceled && Interlocked.Exchange(ref _userCanceled, 0) == 0);
+
+            return new ValueTask<ReadResult>(readResult);
+        }
+
+        private async ValueTask<ReadResult> ReadAsyncInternalAwaited(ValueTask<ReadResult> readTask, CancellationToken cancellationToken = default)
+        {
+            var readResult = await readTask;
+
+            // Ignore the canceled readResult unless it was canceled by the user.
+            while (readResult.IsCanceled && Interlocked.Exchange(ref _userCanceled, 0) == 0)
+            {
+                readResult = await _context.Input.ReadAsync(cancellationToken);
+            }
+
+            return readResult;
```

Comment:
@stephentoub Is this acceptable ValueTask usage?

I think we're okay because this always either awaits the ValueTask or calls GetAwaiter().GetResult(). And it only calls GetAwaiter().GetResult() if we've verified ValueTask.IsCompletedSuccessfully. I know we don't pool ValueTasks yet, but I don't want to introduce more bad code like what you highlighted in #16876.

Reply:
We should turn on the analyzer

Reply:
This looks fine.  And the analyzer shouldn't warn about it; if it does,  we should fix the analysis.